flow/record/base.py

```diff
@@ -662,7 +689,56 @@ def DynamicDescriptor(name, fields):
     return RecordDescriptor(name, [("dynamic", field) for field in fields])
 
 
-def open_path(path, mode, clobber=True):
+def open_stream(fp: BinaryIO, mode: str) -> BinaryIO:
+    if not hasattr(fp, "peek"):
+        fp = Peekable(fp)
+
+    # We peek into the file at the maximum possible length we might need, which is the amount of bytes needed to
+    # determine whether a stream is a RECORDSTREAM or not.
+    peek_data = fp.peek(RECORDSTREAM_MAGIC_DEPTH)
+
+    # If the data stream is compressed, we wrap the file pointer in a reader that can decompress accordingly.
+    if peek_data[:2] == GZIP_MAGIC:
+        fp = gzip.GzipFile(fileobj=fp, mode=mode)
+    elif HAS_BZ2 and peek_data[:3] == BZ2_MAGIC:
+        fp = bz2.BZ2File(fp, mode=mode)
+    elif HAS_LZ4 and peek_data[:4] == LZ4_MAGIC:
+        fp = lz4.open(fp, mode=mode)
+    elif HAS_ZSTD and peek_data[:4] == ZSTD_MAGIC:
+        dctx = zstd.ZstdDecompressor()
+        fp = dctx.stream_reader(fp)
+
+    return fp
+
+
+def find_adapter_for_stream(fp: BinaryIO) -> tuple[Peekable, str]:
+    # We need to peek into the stream to be able to determine which adapter is needed. The fp given to this function
+    # might already be an instance of the 'Peekable' class, but might also be a different file pointer, for example
+    # a transparent decompressor. As calling peek() twice on the same peekable is not allowed, we wrap the fp into
+    # a Peekable again, so that we are able to determine the correct adapter.
+    fp = Peekable(fp)
+    peek_data = fp.peek(RECORDSTREAM_MAGIC_DEPTH)
+    if peek_data[:3] == AVRO_MAGIC and HAS_AVRO:
```

Comment:
Move HAS_AVRO as first check

```suggestion
    if HAS_AVRO and peek_data[:3] == AVRO_MAGIC:
```